### ISSUE TITLE
Nesting Block Elements

### ISSUE THREAD

**Issue by jayaike** [ianstormtaylor/slate#5358]
**Problem**

I am creating a `FeedbackElement` that can wrap any element. So it could appear at the root of the editor (ie. one of the editors children) or as a child of some block node: like a paragraph. `FeedbackElement`s can contain other `FeedbackElements`, or really any element, like images, lists, etc.

The only issue with this is that the default `normalizeNode` does not permit this (especially as a child of a paragraph). I could get this to work by completely disabling normalisation using a plugin that modifies `normalizeNode` to a noop function. But, of course, you don't want to do this because it creates its own sets of bugs.

The type definition for feedback is as follows: 

```
export type FeedbackElement = {
  type: "feedback";
  feedback: string;
  children: Element[];
};
```

**Solution**
There should be a way to nest block elements.

[
    {
        "type": "paragraph",
        "children": [
            {
                "text": "example "
            },
            {
                "type": "feedback",
                "children": [
                     {
                          "text": "example "
                    },
                ]
            },
        ]
    },
    {
        "type": "paragraph",
        "children": [
            {
                "text": ""
            }
        ]
    }
]

**Comment by tiberiuichim** [comment]
@jayaike I've been exploring this similar problem, here's possible fixes:

- override `normalizeNode`, wrap the inline children with a container `<span>` or something similar. This way Slate sees element as first child
- my chosen fix was to write code in my `normalizeNode` that lifts the child `<ul>` to be a sibling of its original parent `<li>`

**Comment by jayaike** [comment]
@tiberiuichim yea im also thinking of modifying the default normalizeNode fn to support my needs. 

- Allow merging of adjacent text nodes that have the same properties
- Allow nesting of block nodes

I see no reason why block nodes should not be nested. Nested block nodes would be similar to inline-block in regular HTML which is of course a thing.

**Comment by tiberiuichim** [comment]
@jayaike We usually implement isInline => true for all our custom elements.

**Comment by jayaike** [comment]
yes i considered this but that wont cover the case (for me) of where i wrap maybe 4 block elements. since inline elements dont typically contain block elements.

whereas a block element can contain block elements and inline elements.

my use case would be something like this.
```
<feedback>
    <p>Here is some random <feedback>data</feedback></p>
   <feedback><img src="..." /></feedback>
</feedback>
```

**Comment by allison-luminos** [comment]
Does anyone have an example of how to overwrite the normalizeNode method to support this?

**Comment by qj10aw** [comment]
@12joan So overriding `normalizeNode` will make the editor vernerable to pasting? I'm trying to override becuase of the pasting, when the source's content is something like this: 

From this nested structure
```
<ul>
  <li>bullet
    <ol>
      <li>One</li>
      <li>Two</li>
      <li>Three</li>
    </ol>
  </li>
  <li>more bullet</li>
</ul>
```

(1) Desired structure
```
<ul>
  <li>bullet</li>
</ul>
<ol>
  <li>One</li>
  <li>Two</li>
  <li>Three</li>
</ol>
<ul>
  <li>more bullet</li>
</ul>
```

(2) But this would do as well
```
<ul>
  <li>bullet</li>
  <li>One</li>
  <li>Two</li>
  <li>Three</li>  
  <li>more bullet</li>
</ul>
```

Probably making (2) wouldn't be too problematic with overriding `normalizeNode?`
I just tried pasting in [platejs.org](https://platejs.org/) but like you mentioned, everything seems to be pasted as just plain <p>
This could be make things simpler but it's a big counterintuitive for users?

**Comment by qj10aw** [comment]
@12joan Thanks! I added `Tranforms.liftNodes` to the `normalizeNode` override like this: 

```
const withCustomPlugins = (editor: CustomEditor) => {
  const { normalizeNode } = editor;

  editor.normalizeNode = (entry) => {
    const [node, path] = entry;

    const isNestedList = 
      Element.isElement(node) &&
      (node.type === 'bulleted-list' || node.type === 'bulleted-list') &&
      Element.isElement(Node.parent(editor, path)) &&
      ((Node.parent(editor, path) as Element).type === 'bulleted-list' ||  // condition (1)
        (Node.parent(editor, path) as Element).type === 'numbered-list' || // condition (2)
        (Node.parent(editor, path) as Element).type === 'list-item'); // condition (3)

    if (isNestedList) {
      Transforms.liftNodes(editor, { at: path });
      return;
    }

    normalizeNode(entry);
  }

  return editor;
}
```

It mostly works but the sibling text node of the lifted node is removed when the nesting is more than two level deeper. 
For example, with this nested html 

<ul>
  <li>One
  	<ul>
      <li>Two
       <ul>Three</ul>
      </li>
      <li>Four</li>
    </ul>    
  </li>
  <li>Five</li>
</ul>

The fragement before normalizing was like this: 
```
[
  {
    type: 'bulleted-list',
    children: [
      {
        type: 'list-item',
        children: [
          { text: 'One' },
          {
            type: 'bulleted-list',
            children: [
              {
                type: 'list-item',
                children: [
                  { text: 'Two' },
                  {
                    type: 'bulleted-list',
                    children: [
                      {
                        type: 'list-item',
                        children: [{ text: 'Three' }],
                      }
                    ]
                  }
                ]
              },
              {
                type: 'list-item',
                children: [
                  { text: 'Four' }
                ]
              }
            ]
          }
        ]
      },
      {
        type: 'list-item', 
        children: [{ text: 'Five' }]
      }
    ]
  }
]
```

But after normalizing, everything's top level node which is good but 'Two' is gone

Resulted:

```
[
  {
    type: 'bulleted-list',
    children: [{ type: 'list-item', children: [{ text: 'One' }] }]
  },
  {
    type: 'bulleted-list',
    children: [{ type: 'list-item', children: [{ text: 'Three' }] }]
  },
  {
    type: 'bulleted-list',
    children: [{ type: 'list-item', children: [{ text: 'Four' }] }]
  },
  {
    type: 'bulleted-list',
    children: [{ type: 'list-item', children: [{ text: 'Five' }] }]
  },
]
```

I'm looking into the API doc how to correctly target the lifting path for UL/OL under LI, in order not to remove it's sibling text,... but I'd appreciate any advice. For the implementation, I'm not making anything inline except for link element, and my settings are almost the same with the paste HTML [example](https://www.slatejs.org/examples/paste-html)

**Comment by 12joan** [comment]
@qj10aw Please could you create a code sandbox or an example in [Slate Transform Explorer](https://5mxwx9.csb.app/) and post your question on Slack?
